lib/Sema/SemaBounds.cpp

```diff
@@ -1418,20 +1434,280 @@ namespace {
+      }
+
+      void SetUpper(llvm::APSInt &Upper) {
+        UpperOffset = Upper;
+      }
+
+      void Dump(raw_ostream &OS) {
+        OS << "ConstantRange:\n";
+        OS << "Base: ";
+        if (Base)
+          Base->dump(OS);
+        else
+          OS << "nullptr\n";
+        SmallString<12> Str1;
+        SmallString<12> Str2;
+        LowerOffset.toString(Str1);
+        UpperOffset.toString(Str2);
+        OS << "Lower offset:" << Str1 << "\nUpper offset:" << Str2 << "\n";
+      }
+    };
+
+    // Convert an expression E into a base and offset form.
+    // - If E is pointer arithmetic involving addition or subtraction of a
+    //   constant integer, return the base and offset.
+    // - Otherwise set base to E and offset to 0.
+    // TODO: we use signed integers to represent the result of the Offset.
+    // We can't represent unsigned offsets larger the the maximum signed
+    // integer that will fit pointer width.
+    void SplitIntoBaseAndOffset(const Expr *E, unsigned PointerWidth,
+                                const Expr *&Base, llvm::APSInt &Offset) {
+      if (const BinaryOperator *BO = dyn_cast<BinaryOperator>(E->IgnoreParens())) {
+        if (BO->isAdditiveOp()) {
+          Expr *Other = nullptr;
+          if (BO->getLHS()->getType()->isPointerType()) {
+            Base = BO->getLHS();
+            Other = BO->getRHS();
+          } else if (BO->getRHS()->getType()->isPointerType()) {
+            Base = BO->getRHS();
+            Other = BO->getLHS();
+          } else
+            goto exit;
+          assert(Other->getType()->isIntegerType());
+          if (Other->isIntegerConstantExpr(Offset, S.Context)) {
+            // Widen the integer to the number of bits in a pointer.
+            if (Offset.getBitWidth() > PointerWidth)
+              goto exit;
+            else if (Offset.getBitWidth() < PointerWidth)
+              Offset = Offset.extend(PointerWidth);
+            // If the offset is an unsigned integer, convert it to a signed integer.
+            if (Offset.isUnsigned()) {
+              if (Offset > llvm::APSInt(Offset.getSignedMaxValue(PointerWidth)))
+                goto exit;
+              Offset = llvm::APSInt(Offset, false);
+            }
+            // Normalize the operation by negating the offset if necessary.
+            if (BO->getOpcode() == BO_Sub)
+              Offset = llvm::APSInt(PointerWidth, false) - Offset;
+            uint64_t ElemBitSize = S.Context.getTypeSize(Base->getType()->getPointeeOrArrayElementType());
+            uint64_t ElemSize = S.Context.toCharUnitsFromBits(ElemBitSize).getQuantity();
+            llvm::APSInt ElemInt = llvm::APSInt(llvm::APInt(PointerWidth, ElemSize), false);
+            bool Overflow;
+            Offset = Offset.smul_ov(ElemInt, Overflow);
+            if (Overflow)
+              goto exit;
+            return;
+          }
+        }
+      }
+
+    exit:
+      Base = Base = E->IgnoreParens();
```

Comment:
Not clear what the double assignment to `Base` is doing here.

Reply:
Typo - I'll fix this.